Transitively add pinned dirs

mosteo opened the pull request:
When a pin to a local crate is added, recursively add its pins to the solution too.

Goes on top of #711.

@onox, if you haven't yet tested #711, you can test this one directly for the use case we talked about on gitter.

Testcase incoming.

mosteo committed:
Store relative pins when given as such

mosteo committed:
Ensure pinned dirs use portable paths when relative

mosteo committed:
Adjustments in testsuite for new relative links

mosteo committed:
Recursively add links when linking crates

mosteo committed:
Compound relative paths when adding links

mosteo committed:
Restore inadvertently removed comment

onox commented:
If I add orka in orka_glfw then `alr` detects orka's dependencies, but I get a failed precondition after proceeding:

```
detail: Narrowing down dependency orka* as orka~0.2.0
detail: Using default value for configuration 'warning.caret': 'TRUE'
detail: Using default value for configuration 'warning.caret': 'TRUE'
info: Requested changes:
info: 
info:    ✓ orka ~0.2.0 (add)
info: 
info: Changes to dependency solution:
info: 
info:    =  libglfw3        3.2.1 (unchanged)                 
info:    +⊙ orka            0.2.0 (new,pin=../orka)           
info:    +⊙ orka_egl        1.0.0 (new,pin=../orka_egl)       
info:    +⊙ orka_simd       1.0.0 (new,pin=../orka_simd)      
info:    +⊙ orka_transforms 1.0.0 (new,pin=../orka_transforms)
info:    +⊙ orka_types      1.0.0 (new,pin=../orka_types)     
info: 
Do you want to proceed?
[Y] Yes  [N] No  (default is Yes) yes
detail: Manifest updated, fetching dependencies now
error: failed precondition from alire-solutions.ads:313
```

onox commented:
I guess there is at the moment no way to regenerate a lock file with a command (because the paths are stored in the lock file)? I had to remove the dependencies from the toml file and then re-add them to recreate the lock file.

mosteo commented:
Thanks for the report, @onox. I'll try to reproduce.

> I guess there is at the moment no way to regenerate a lock file with a command (because the paths are stored in the lock file)? I had to remove the dependencies from the toml file and then re-add them to recreate the lock file.

`alr update` should rebuild the lockfile, taking into account pre-existing pins. In this case with the exception, I'm not sure yet what would happen.

mosteo committed:
Relax too restrictive checks

In the case of linked releases, it doesn't matter if these contain expressions
as we don't write them to disk. As for the spurious warning during loading of
links, we may be transitively loading and the assumption on the current
directory is faulty in that case. The check must be performed elsewhere.

mosteo commented:
@onox, I could reproduce your problem using your repository. That's a lot of crates you have there :D. It should work after my last push.

mosteo committed:
Require approval for links to non-existent dirs

mosteo committed:
Show broken links in dependency image

mosteo committed:
Testsuite: new test and minimal tweaks

mosteo committed:
Merge remote-tracking branch 'alire/master' into feat/transitive-links

onox commented:
Thanks, the failed pre-condition is now gone.

I notice that `alr` adds the relative paths twice (once as a relative path and once as an absolute path):

```
detail: Env: crates appends '../orka' to 'GPR_PROJECT_PATH'
detail: Env: crates appends '/absolute/path/to/orka' to 'GPR_PROJECT_PATH'
```

Fabien-Chouteau commented:
@onox it looks to me like you are making you life harder having so many crates.
What is the reason to have `orka-types`, `orka-smid` and `orka-transform` outside of `orka`?

onox commented:
Because `orka-simd` and `orka-transforms` provide bindings for SSE/AVX and vector/matrix operations. I also use these for a numerical integrator for physics. These two small crates are also perfectly reusable by other projects that do not need or want to depend on the OpenGL and rendering stuff.

Anyway, I've fixed the build problems I was having.

Fabien-Chouteau commented on src/alire/alire-solutions.adb:
```diff
@@ -287,6 +289,53 @@ package body Alire.Solutions is
+      do
+         if Linked_Root.Is_Valid and then Linked_Root.Value.Has_Lockfile then
+            for Dep of Linked_Root.Value.Solution.Links loop
+               Result := Result
```
This statement is not very readable in my opinion.

Maybe the result of `Externals.Softlinks.New_Softlink` could be in constant.
same for `Linked_Root.Value.Solution.State`.

mosteo replied:
I have refactored things a bit for clarity.

Fabien-Chouteau commented on src/alire/alire-utils-user_input.adb:
```diff
@@ -25,6 +27,28 @@ package body Alire.Utils.User_Input is
        when Always => "always");
 
+   ------------------
+   -- Approve_Path --
```
Discrepancy with actual function name.

mosteo replied:
Thanks, fixed.

mosteo commented:
> I notice that `alr` adds the relative paths twice (once as a relative path and once as an absolute path):

Thanks for catching that.

> Anyway, I've fixed the build problems I was having.

So these weren't caused by alr?

mosteo committed:
Code review fixes

mosteo committed:
Softlinks: report absolute project paths

onox commented:
> So these weren't caused by alr?

Correct, not caused by alr.

pmunts commented:
Argh...another GNAT internal error.  Abandoning for now.
